### PR TITLE
fix vlm support asyncgrpo

### PR DESCRIPTION
**Pull request by AmineDiro**
# What does this PR do?

Scoped down from #6515 on purpose

Adds support to VLMs for `AsyncGRPOTrainer`, this DOESNT support multimodal training. 

- Load through `create_model_from_path`
- **Freeze everything outside the text tower**: no images in the dataset means the vision tower is never exercised.
- Fix: `compute_flops_per_token` and `patch_chunked_lm_head` 
- Strip `_checkpoint_wrapped_module.` from synced weight names, matching sync GRPO's `_fix_param_name_to_vllm`.

cc @adithya-s-k  @qgallouedec 

---

> [!NOTE]
> **Medium Risk**
> Touches async training load path, weight sync to vLLM, and log-prob/MFU plumbing for VLMs; multimodal and hybrid-model caveats remain documented but not fully mitigated in code.
> 
> **Overview**
> **AsyncGRPOTrainer** can now load VLMs (e.g. Qwen3-VL) via `create_model_from_path` and run **text-only** GRPO: the vision stack stays loaded for vLLM name parity but is **frozen** (no grads, no optimizer state, omitted from NCCL weight sync). Trainable weights are limited to the text submodule and output embeddings.
> 
> Shared fixes for composite configs: **`patch_chunked_lm_head`** and **`compute_flops_per_token`** read hyperparameters from `model.config.get_text_config()` instead of the top-level config. Weight metadata and streaming skip `requires_grad=False` params and normalize names with `_checkpoint_wrapped_module.` removal (aligned with sync GRPO).
> 
> Docs add a **Vision-language models** section (serve example, frozen-tower behavior, and warnings: no image prompts yet, hybrid Qwen3.5/3.6 need `flash-linear-attention`). New **`TestAsyncGRPOTrainerVLM`** covers freeze, sync scope, and a short train step on tiny Qwen VL checkpoints.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 60544e4989d7a9f4d40c3a3732cd8fe2eeeb4a27. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: docs/source/async_grpo_trainer.md [hunk 1]
```diff
@@ -78,6 +78,25 @@ CUDA_VISIBLE_DEVICES=0 VLLM_SERVER_DEV_MODE=1 vllm serve Qwen/Qwen3-4B \
 CUDA_VISIBLE_DEVICES=1 accelerate launch train_async_grpo.py
 ```
 
+## Vision-language models
+
+Checkpoints that ship a vision tower — Qwen3.5, Qwen3.6, Qwen3-VL, … — can be trained on **text-only** datasets. Pass the model id as usual; nothing else changes:
```
Comment (qgallouedec) [inline docs/source/async_grpo_trainer.md]: "Checkpoints that ship a vision tower — Qwen3.5, Qwen3.6, Qwen3-VL, … —"
in other words: Vision language models ;)

File: docs/source/async_grpo_trainer.md [hunk 2]
```diff
@@ -78,6 +78,25 @@ CUDA_VISIBLE_DEVICES=0 VLLM_SERVER_DEV_MODE=1 vllm serve Qwen/Qwen3-4B \
...
+```bash
+CUDA_VISIBLE_DEVICES=0 VLLM_SERVER_DEV_MODE=1 vllm serve Qwen/Qwen3.5-2B \
+    --max-model-len 4096 \
+    --logprobs-mode processed_logprobs \
+    --weight-transfer-config '{"backend":"nccl"}'
+```
+
+Two things are worth knowing about how this works:
+
+- **The whole model is loaded, not just its text tower.** vLLM serves the `*ForConditionalGeneration` architecture, whose parameters are named `model.language_model.*` and `model.visual.*`. Loading only the text tower would name them `model.*` and every NCCL weight transfer would fail on unknown keys. So the vision tower is loaded too, and costs GPU memory for its weights.
```
Comment (qgallouedec) [inline docs/source/async_grpo_trainer.md]: > Loading only the text tower would name them `model.*` and every NCCL weight transfer would fail on unknown keys. So the vision tower is loaded too, and costs GPU memory for its weights.

I think this is an internal detail, it's not worth explaining in the doc. But NIT

File: docs/source/async_grpo_trainer.md [hunk 3]
```diff
@@ -78,6 +78,25 @@ CUDA_VISIBLE_DEVICES=0 VLLM_SERVER_DEV_MODE=1 vllm serve Qwen/Qwen3-4B \
...
+- **The vision tower is frozen.** A text-only dataset never produces image tokens, so the tower is never exercised by the forward pass. Everything outside the text tower (vision tower, multimodal projector) has `requires_grad=False`: it gets no gradients and no optimizer state, and weight sync skips it entirely — the server keeps the values it loaded from the checkpoint.
```
Comment (qgallouedec) [inline docs/source/async_grpo_trainer.md]: good!

File: tests/experimental/test_async_grpo_trainer.py [hunk 4]
```diff
@@ -300,6 +300,88 @@ def reset(self, **kwargs): ...
...
+            reward_funcs=dummy_reward_func,  # unused: the stub pre-computes rewards, but the trainer requires it
+            args=training_args,
+            train_dataset=dataset,
+            rollout_worker=_StubRolloutWorker(AutoTokenizer.from_pretrained(model_id), dataset, num_generations=3),
+            weight_transfer=_StubWeightTransfer(),
+        )
+
+    def test_vision_tower_is_frozen(self, model_id):
+        trainer = self._trainer(model_id)
+        assert trainer._is_vlm
```
Comment (qgallouedec) [inline tests/experimental/test_async_grpo_trainer.py]:
```suggestion
```

let's not test on private vars.

File: tests/experimental/test_async_grpo_trainer.py [hunk 5]
```diff
@@ -300,6 +300,88 @@ def reset(self, **kwargs): ...
...
+        trainer.train()
+
+        assert trainer.state.log_history[-1]["train_loss"] is not None
+
+        for n, param in previous_params.items():
+            new_param = trainer.model.get_parameter(n)
+            if new_param.requires_grad:
+                assert not torch.equal(param, new_param), f"Parameter {n} has not changed."
+            else:
+                assert torch.equal(param, new_param), f"Frozen parameter {n} has changed."
```
Comment (qgallouedec) [inline tests/experimental/test_async_grpo_trainer.py]: If there is a simple way to do this, I think it would be better to branch on whether the param in in the vision tower or not.

File: trl/experimental/async_grpo/async_grpo_trainer.py [hunk 6]
```diff
@@ -821,6 +824,11 @@ def __init__(
         self.aux_loss_enabled = is_moe and args.router_aux_loss_coef != 0.0
         self.router_aux_loss_coef = args.router_aux_loss_coef
 
+        self._is_vlm = text_config is not model.config
+        if self._is_vlm:
+            model.model.requires_grad_(False)
+            model.model.language_model.requires_grad_(True)
```
Comment (qgallouedec) [inline trl/experimental/async_grpo/async_grpo_trainer.py]: for ref, non-blocking, this might not work in general, for ref, this is another approach https://github.com/huggingface/trl/pull/6618/changes#diff-5f63a484ac9f57663f1e8d18ab41afe1d55e7909133191175cbc4943197f39adR132-R146

**Review by qgallouedec — APPROVED** [review]
Thanks, I didn't test myself, but overall it looks good, just some minor non-blocking remarks

**Review by qgallouedec — CHANGES_REQUESTED** [review]
Actually the CI fails.

File: tests/experimental/test_async_grpo_trainer.py (hunk 5 above)
Comment (AmineDiro) [reply tests/experimental/test_async_grpo_trainer.py]: oh you mean to make sure the param comes from the vision tower by module name for example ?

File: tests/experimental/test_async_grpo_trainer.py [hunk 7]
```diff
@@ -304,6 +304,109 @@ def reset(self, **kwargs): ...
...
+        assert streamed
+        assert streamed.keys() == {n for n, p in trainer.model.named_parameters() if p.requires_grad}
+        assert any(name.startswith("model.language_model.") for name in streamed)
+        assert not any(".visual." in name or ".vision_tower." in name for name in streamed)
+
+    def test_train(self, model_id):
+        if "Moe" in model_id:
+            # `compute_flops_per_token` reads `num_local_experts`, `intermediate_size` and `decoder_sparse_step`,
+            # none of which exist on `Qwen3_5MoeTextConfig`. Text-only Qwen3.5-MoE hits this too.
+            pytest.skip("compute_flops_per_token does not support the Qwen3.5-MoE config shape")
```
Comment (qgallouedec) [inline tests/experimental/test_async_grpo_trainer.py:393]: Ideally, you want to open an issue and put the number in the skip reason

File: trl/experimental/async_grpo/async_grpo_trainer.py [hunk 8]
```diff
@@ -1255,7 +1274,7 @@ def _log_step_metrics(self) -> None:
         ## `_wall_clock` divides by `perf/step_s`: the whole step, rollout waits included  and says what fraction of the allocation actually became training.
         if self._step_forward_tokens > 0:
             mean_seq_len = self._step_seq_len_weighted / self._step_forward_tokens
-            flops_per_token = compute_flops_per_token(self.model.config, int(mean_seq_len))
+            flops_per_token = compute_flops_per_token(self.model.config.get_text_config(), int(mean_seq_len))
```
Comment (qgallouedec) [inline trl/experimental/async_grpo/async_grpo_trainer.py:1278]: nice, can you do the same fix in https://github.com/huggingface/trl/blob/12b43d00a5bf2e121294babab9d3d0327e276363/trl/experimental/async_distillation/async_distillation_trainer.py#L1412

**Review by qgallouedec — APPROVED** [review]
nice! two nits, but you can merge w/o addressing, your call
